compiler/rustc_const_eval/src/interpret/memory.rs

```diff
@@ -1222,6 +1222,76 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
 
         Ok(())
     }
+
+    pub fn swap_memory(
+        &mut self,
+        x_ptr: Pointer<Option<M::Provenance>>,
+        y_ptr: Pointer<Option<M::Provenance>>,
+        count: u64,
+        layout: ty::layout::TyAndLayout<'tcx>,
+    ) -> InterpResult<'tcx> {
+        let size = layout.size;
+        let align = layout.align.abi;
+        let Some((x_alloc_id, x_offset, _)) = self.get_ptr_access(x_ptr, size, align)? else {
+            // Called on ZST so it is noop.
+            return Ok(())
```

Comment:
This is not good. If the 2nd pointer is invalid you would miss the UB! It is important to check all conditions before short-circuiting. Roughly follow that `mem_copy` above does.

Reply:
Hi, @RalfJung !
Is is now correct way or do I need to call `M::before_memory_read` and `M::before_memory_write` too?

Reply:
You'll need to call these hooks the same way `mem_copy` does.